### ISSUE TITLE
Final assessment of InputDefinition/OutputDefinition.metadata

### ISSUE THREAD

**Issue by sryza** [dagster-io/dagster#3427]
Once we add a metadata field to InputDefinition and OutputDefinition, it will be hard to go back.  This issue is to track taking a final look at this before the 0.10.0 release.

Some other potential options:
* tags, which would provide less hierarchy and more structure
* config

**Comment by yuhan** [comment]
TLDR: I'd suggest to remove metadata and only have input/output_config

Long answer:
I think metadata on definitions will be mode-independent. When writing a pipeline, my initial answer to where to put the per-step info was to put info like table_name or slack's channel name on InputDefinition/OutputDefinition.metadata. But when I was testing it in different modes, i.e. unittest, dev, prod, I found that I certainly want the capability to switch metadata between modes.

One could argue that in some cases, where the user never expects the metadata to change, InputDefinition/OutputDefinition.metadata is convenient: because it's tied to the solid definition
1) it's easy to understand the business logic
2) when you don't want users to worry about specify metadata with configs every time the solid is reused. - which I think config schema's default value could address this concern too.

But I think using the input/output_config_schema is more flexible and consistent:
1) it's mode-dependent
2) config schema is well structured as opposed to the current metadata freeform - when I was reusing a solid, it took me quite a while to figure out what the fields in the metadata are and which of them are required, but config schema can clearly address this user experience
3) keeping both metadata and config, it's hard to explain to users when to use one over the other. I'd prefer to use input config only to avoid the ambiguity.

**Comment by sryza** [comment]
Something to consider is whether we can get close to metadata with `configured`.  E.g.
```
@solid(output_defs=[OutputDefinition().configured({"table_name": "users"})])
def make_users_table():
    ...

@object_manager(output_config_schema={"table_name": str})
def snowflake_object_manager():
    ...

@object_manager(output_config_schema={"table_name": str})
def postgres_object_manager():
    ...
```

What's a little weird here is that, at the time `configured` is called, we don't know what the config schema for the output will be.  However, at least with our current implementation, we don't need that config schema at that time.  Configuration passed to 'configured` is applied at environment config construction time.

**Comment by sryza** [comment]
Another potential direction: expect users to use dagster types for encoding this information.  E.g.

```
@solid(output_defs=[OutputDefinition(DagsterTypeWithMetadata(name="users", metadata={"table_name": "users"})])
def make_users_table(_):
    ...

class MyObjectManager(ObjectManager):
    def handle_output(context, obj):
        table_name = context.dagster_type.metadata["table_name"]
        save_object_to_table(obj, table_name)
```

**Comment by schrockn** [comment]
Definitely feel like we should not encode this on type. You should be able to construct a data lake graph with vanilla dataframe types IMO. This would require creating N types for the sole purpose of metadata.

**Comment by sryza** [comment]
> This would require creating N types for the sole purpose of metadata.

I'm not particularly convinced this is a good idea, but we could potentially address it with some sort of wrapper:

```
UsersType = SparkDFType("users", cols=[...])

@solid(output_defs=[
    OutputDefinition(DagsterTypeWithMetadata(UsersType, metadata={"table_name": "users"})
])
def make_users_table(_):
    ...
```

**Comment by schrockn** [comment]
Just doesn't make too much sense to me to tie input and output managers to input/outputs and the metadata they use to types

**Comment by sryza** [comment]
@schrockn - do you want to weigh in here?  When we chatted in person, I think you mentioned you had some concerns about moving this to config?

**Comment by schrockn** [comment]
Yes I have serious of concerns about this.

Config is generally not meant to encode information that will never vary between different invocations. This results in repeated blocks of config that never change and that in fact you do not want to vary.

I think this is part of a broader problem in that we have not formulated or articulated a coherent story about what should be encoded in config versus what is encoded on definitions. This leaves us in weird state where we don't have clear guidance on what should be modelled as a "solid factory" versus config. `configured` is very elegant and valuable but in some ways has actually either compounded or further exposed this problem based on how one frames it.

In my opinion, we should be driving towards a world where config is viewed as the layer between application development and operations. We should only encode in config what one would want to credibly alter without changing code or redeploying.

**Comment by sryza** [comment]
I buy that.  What you're articulating strongly matches some of my initial reservations about the system.
